lib/BinaryNode.cpp

```diff
@@ -44,12 +44,19 @@ BinaryNode::~BinaryNode() {
 }
 
 void BinaryNode::clear() {
-   if (len) {
-      assert(ptr);
-      free(ptr);
-      len = 0;
-      ptr = 0;
-   }
+    // issue #2982: must check 'ptr', len may be 0 with memory allocated
+    if (ptr) {
+        free(ptr);
+        if (len) {
+            len = 0;
+        }
+        ptr = nullptr;
+    }
+#ifdef DEBUG
+    else {
+        assert(!len);
+    }
+#endif
```

Comment:
yes for all of your comments, but if you want the best performance, you will not update memory when it doesn't need it
that's why you see code like:
```
if (x) {
    x = 0;
}
```
In qore - it's a pure performance optimization

Reply:
and are u sure that testing the memory will be that faster then writing to it? also - the write then (probably minor case) will take longer (testing + writing)

Reply:
the idea is that if you just compare that it loads the CPU cache in the worst case, but if you also write, that it will end up flushing the cache as well, so the idea is to avoid writes when possible: https://stackoverflow.com/questions/910423/which-is-faster-comparison-or-assignment

I have not done real-world testing on this, but my understanding is that is was an accepted performance optimization for the linux kernel, and therefore could make qore just a very very tiny bit faster